WIP: resize textures using GL_PROXY mechanism (and other things)

illwieckz opened the pull request:
1. Resize textures using GL_PROXY mechanism
2. Introduce picMin and picMax

The ATI X1950 PRO can display the game with medium preset including full size textures except the skybox that is `1024^6` size:

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-27_184412_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-27_184412_000.jpg)

With `r_picMax` set to `512`, everything looks good:

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-27_184517_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-27_184517_000.jpg)

The `r_picMax` cvar expects the size of an edge, so `r_picMax` 512 will scale down every image larger than a square image of `512×512` size.

Using `r_picMax 512` will scale down textures from `tex-vega` that are `1024×1024` large but will not scale down textures from `tex-pk01` that are `512x512` large. This way a map using both vega and pk01 textures will look highly textured with `512×512` textures everywhere while using `r_picmip 1` would have displayed a mix of nice `512x512` and ugly `256×256` images.

Both `r_picmip` and `r_picMax` can be used at the same time. For example, by using `r_picmip 1` and `r_picMax 512`, once minimap is set as nopicmip (see https://github.com/Unvanquished/Unvanquished/pull/1230 ), the game will scale down all textures but the minimap to keep it clear, unless it is larger than 512, _but minimaps with size up to 512 will not be scaled down_…

I added two shader keywords: `picMin` and `picMax`.

The `picMin` one allows an artist to tell the renderer to never downscale a texture under this size. This is useful for models for example, since seeing details like the eyes and others is better.

The `picMax` allows an artist to tell the renderer to downscale the texture to this size, even if the player is playing full size.

I added another cvar, `r_picMin`, which allows to override the `picMin` value set in shader, and when it is set to `-1`, never downscale images that has `picMin` value (it would be like making them `nopicmip`).

slipher commented:
It looks like there is some code right next to the code you added which is supposed to do exactly the same thing. What is `glConfig2.maxCubeMapTextureSize` on your test machine?

illwieckz commented:
I get `4096` on the for `glConfig2.maxCubeMapTextureSize` on the ATI X1950 PRO.

illwieckz commented:
Interestingly those are the values on the original code to scale down lightmaps:

```
scaledWidth: 1024, scaledHeight: 1024, glConfig2.maxCubeMapTextureSize: 4096
```

Is `glConfig2.maxCubeMapTextureSize` the size of a single face or the size of the whole cube? If it's the size of a single face then the driver is buggy, if it's the size of the whole cube then we have to compare `scaledWidth` and `scaledHeight` with `glConfig2.maxCubeMapTextureSize / 6` instead.

That said, this hardware is not meant to support `4096` sized standard flat textures, so it's probably the summed size of all the faces of the skybox.

Testing against `glConfig2.maxCubeMapTextureSize / 6` fixes the issue.

I still think the `r_picMax` feature is good even if unneeded to fix bugs. For example setting `r_picMax` to `512` still produces high definition rendering and it's better than doing `r_picmip 1` that would reduce `4096` sided large textures to still-heavy `2048` ones and ok-ish `512` sided textures to ugly `256` ones.

Basically `r_picmip` looks a relic from a past where all textures had the same size and were low definition anyway, and then user would somewhat expect the resulting images having the same size. To me this `r_picMax` feature is more useful than `r_picmip`, this is what I would naturally look for instead.

illwieckz commented:
I added a commit to fix skybox scaling according to the hardware limits.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -865,7 +865,7 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips,
 	// deal with a half mip resampling
 	if ( image->type == GL_TEXTURE_CUBE_MAP )
 	{
```
This is the existing code I was talking about. Instead of creating yet another downscaling mechanism, I suggest using this one by, say, taking the min of picmax and the glconfig's max size value.
Although I think the existing code is broken for compressed textures, since it doesn't set `picmip`.

illwieckz replied:
glconfig max size value is not used anymore.

slipher reviewed (commented):
I doubt you are supposed to divide the cubemap max size by 6, but who knows. According to [this](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glGetError.xhtml) the max size is supposed to be at least 1024, so reporting the correct value would be illegal... I wonder if 4096 is a default value when the implementation doesn't provide a max size.

I just noticed something really stupid... GL error checking is disabled by default (`r_ignoreGLErrors 1`). Turning it on, however, terminates the program the first time as an error. Of course the sensible thing to do would be a non-fatal log message. This should help diagnose things, since using a texture larger than the max size is one of the GL errors which can be produced.

I agree `r_picmip` is a puzzling feature. I guess you could use it to test the downsampling code, or as a brute force way to reduce GPU memory. Both of which could be done with picmax as well. A quick google suggests it has been mostly popular as a cheat to make enemies more visible in various games.

illwieckz commented:
I rewrote entirely the code to compute an image size, or to be more precise, to compute a pixel size in memory (see #378).

I added two shader keywords, `picMin` and `picMax`.

The `picMin` one allows an artist to tell the renderer to never downscale a texture under this size. This is useful for models for example, since seeing details like the eyes and others is better.

The `picMax` allows an artist to tell the renderer to downscale the texture to this size, even if the player is playing full size. This can be useful for example if a texture is shipped in `4096×4096` size but is only displayed on a tiny surface. This allow to ship full size textures even if they are used downscaled in game. It's better for conservancy.

I added another cvar, `r_picMin`, which allows to override the `picMin` value set in shader.

So basically, texture is downscaled using  `picMax` (or `r_picMax` if positive) unless a shader uses `picMin`. If a shader uses `picMin`, the texture is downscaled to the largest value between `picMin` or `picMax` unless `r_picMin` is set. If a shader uses `picMin` and `r_picMin` is set to something greater than zero, then texture is downscaled using `r_picMin`. There is a special value for `r_picMin`: if a shader uses `picMin` and `r_picMin` is set to `-1`, then texture is not downscaled at all. This makes possible to downscale a lot what is known to be downscalable safely, and to not downscale at all what is known to not downscale properly. This is useful for competitive scenes: it makes possible to make world surfaces using only one color but models would be rendered properly.

Pictures may help to understand, in all those screenshots, models shaders have `picMin 128` instruction.

`r_picMax 512` and `r_picMin 0`, every texture is loaded with `512×512` size maximum.

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005229_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005229_000.jpg)

`r_picMax 64` and `r_picMin 0`, every texture is loaded with `64×64` size maximum, except models that are loaded with `128×128` size.

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005252_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005252_000.jpg)

`r_picMax 1` and `r_picMin 0`, every texture is loaded with `1×1` size maximum, except models that are loaded with `128×128` size.

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005313_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005313_000.jpg)

`r_picMax 1` and `r_picMin -1`, every texture is loaded with `1×1` size maximum, except models that are loaded full size.

[![picmax](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005350_000.jpg)](https://dl.illwieckz.net/b/daemon/features/picmax/unvanquished_2020-10-31_005350_000.jpg)

slipher commented:
> The picMax allows an artist to tell the renderer to downscale the texture to this size, even if the player is playing full size. This can be useful for example if a texture is shipped in 4096×4096 size but is only displayed on a tiny surface. This allow to ship full size textures even if they are used downscaled in game.

This doesn't sound like a great practice to encourage, since you would still pay full price in file size and load times. Wouldn't something like this be better as part of the map build process?

illwieckz commented:
This is better as part of a map build process, but a mapper can still write shaders based on third-party texture set and other situations.

Note that I did not added that because of a urge need. I see a real need for picMin so I did it, and right after than I've seen picMax could be done for free. There would be no issue to not have it, but once picMin is implemented, picMax costs nothing, so why not.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -143,6 +143,417 @@ void GL_TextureMode( const char *string )
+int GetFormatPixelSize( GLenum internalFormat )
+{
+
+	int s8 = 8;
```
const / constexpr?

illwieckz replied:
done.

slipher commented on src/engine/renderer/tr_local.h:
```diff
@@ -597,6 +597,7 @@ static inline void halfToFloat( const f16vec4_t in, vec4_t out )
 
 		int            frameUsed; // for texture usage in frame statistics
 
+		int            formatSizeFactor;
```
unused

illwieckz replied:
ah yes, a leftover from work-in-progress implementation. I noticed storing the result of division as an int was bad about precision XD.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -143,6 +143,417 @@ void GL_TextureMode( const char *string )
+			return 8 + 8 + 8;
+		case GL_RGB8_SNORM:
+			return s8 + s8 + s8;
+		case GL_RGB10:
```
Are you sure all these ones which are not divisible by 8 actually take up a non-integer number of bytes?

Probably the engine can't actually load any of the ones like that, so one option would be to just remove ones which are too much work to figure out.

illwieckz replied:
Values taken from https://www.khronos.org/opengl/wiki/GLAPI/glTexImage2D

I think for such situation it's better to be exhaustive even if it's unused: if we start to remove some, it will be hard to know if something is missing and what.

And yes, there is formats with non-integer numbers of bytes. It's even worse with ASTC since pixel size is a float.

illwieckz replied:
Now the code lives in its own file. I've made it as exhaustive as I could like if it was a repurposable library, even if we don't use every format on earth.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -170,286 +580,270 @@ void R_ImageList_f()
 	{
 		image = (image_t*) Com_GrowListElement( &tr.images, i );
 		char buffer[ MAX_TOKEN_CHARS ];
+		char imageType[ MAX_TOKEN_CHARS ];
```
Just use `std::string`

illwieckz replied:
Still `TOTO`.

illwieckz replied:
Hmm, there is still usage of `Com_sprintf`, not sure how to handle that with `std::string`.

slipher replied:
`.c_str()` to get a char*

illwieckz replied:
I still don't know how to do this with `std::string`:

```c
Com_sprintf( imageType, sizeof( imageType ),  "%5i ", image->type );
```

slipher replied:
Com_sprintf -> Str::Format

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -1057,6 +1431,58 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips,
+	> format may be using 64 bits per texel, so its image may have to be 16 times smaller
+	> than an image with the GL_LUMINANCE4 internal format. (Also, images requiring borders
+	> or mipmaps may further reduce the amount of available memory.) 
+	-- http://www.glprogramming.com/red/chapter09.html
```
How do you know which texture format GL_MAX_TEXTURE_SIZE is assuming?

It seems better to use the the method described in the next section, *Texture Proxy*. In that case, it is not necessary to have a database of the number of bits per texel, or to guess at what bit density the maximum value assumes, or whether one should consider the edge length or the total number of pixels.

illwieckz replied:
Things like `GL_LUMINANCE4` that is taken as example and others are not described in current OpenGL wiki, or I did not found them:
https://www.khronos.org/opengl/wiki/GLAPI/glTexImage2D

But we can find some of them in OpenGL 2.1 documentation:
https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glTexImage2D.xml
Or in OpenGL 1.1 documentation:
https://web.archive.org/web/20011231003042/http://www.opengl.org/developers/documentation/Specs/glspec1.1/node79.html

It still don't find formal definition for `GL_RGB`.

I have read people saying GL_RGB is 24 bit, interlaced (`RGBRGBRGB`), like [here](https://forum.openframeworks.cc/t/gl-rgb/72/2). That's not from Khronos though, but that's an example of an answer to solve someone's need to load data directly as `GL_RGB`.

While using `imagelist` command I discovered some of our textures are listed as using `GL_RGBA`, so I may add those formats too:

```
      -w-- -h-- -px- -size------ -mm- -type- -format- -wrap.s- -wrap.t- -name------- 
Debug: Unknown pixel size for undocumented image format 6408 
Debug: Undocumented image format 6408 for image fonts/unifont_0_0_16.png 
  35:  512  256   32      174762 no   2D      6408    s.clmp   t.clmp   fonts/unifont_0_0_16.png 
```

`6408` is `0x1908` wich is `GL_RGBA` in `gl.h`…

So, unless proven wrong,it looks like `GL_RGB` uses 8 bit per pixel, then I see no reason why `GL_RGBA` would not use 8 bit per pixel, but documentation about them is harder to find. Also, unless proven wrong, it looks like `GL_RGB` is the default format.

illwieckz replied:
This code does not exist anymore.

illwieckz commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -143,6 +143,417 @@ void GL_TextureMode( const char *string )
 	}
 }
 
+int GetFormatPixelSize( GLenum internalFormat )
```
In Memory, I can rename it right. Also, it can't return perfect answer in all case, for example it is somewhat wrong with DXT textures when a picture has less than 4 pixels (maybe I can add a comment about that).

illwieckz replied:
Added comments about it.

illwieckz commented on src/engine/renderer/tr_init.cpp:
```diff
@@ -1057,6 +1059,7 @@ ScreenshotCmd screenshotPNGRegistration("screenshotPNG", ssFormat_t::SSF_PNG, "p
 
 		r_picmip = ri.Cvar_Get( "r_picmip", "0",  CVAR_LATCH | CVAR_ARCHIVE );
 		AssertCvarRange( r_picmip, 0, 3, true );
+		r_picMax = ri.Cvar_Get( "r_picMax", "0",  CVAR_LATCH | CVAR_ARCHIVE );
```
I've basically done `max(0, r_picMax)`, and `r_picMin` is meant to accept `-1`. I see no issue with too large values, except they are useless. To me that does not sound enough to require an assert.

illwieckz replied:
Now it does not compare `picMax * picMax` anymore but `picMax` itself, is there still a risk of integer overflow?

illwieckz commented:
I've added a commit temporarily named `wat` that may fix an issue… While reading the original code, I found this part looking weird and that sounded like a typo. I would appreciate opinions on that.

illwieckz commented:
This is basically what does the `wat` commit, before:

```
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 0, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 1, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 2, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 3, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 4, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 5, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 6, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 7, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 8, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 9, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 10, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 11, layer 0
```

after:

```
Debug: Loading compressed image models/buildables/barricade/body_d with 2048×2048 original size, mip 0, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 1024×1024 size, mip 1, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 512×512 size, mip 2, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 256×256 size, mip 3, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 128×128 size, mip 4, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 64×64 size, mip 5, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 32×32 size, mip 6, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 16×16 size, mip 7, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 8×8 size, mip 8, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 4×4 size, mip 9, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 2×2 size, mip 10, layer 0
Debug: Loading compressed image models/buildables/barricade/body_d with 1×1 size, mip 11, layer 0
```

Original behavior sounded wrong, right?

illwieckz commented:
Unfortunately the new code to detect images being too large for the hardware does not work. I never see them being downscaled.

illwieckz commented:
It works, but not the way we expect it for cube maps. If I double width and height when passing them to `CheckImage2D`, the obviously too large size is detected and proper order is given to downscale, and the skybox is rendered properly (because effectively downscaled):

```
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 0
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 1
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 2
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 3
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 4
Debug: Uploading image env/shared_space_src/sky07 with 256×256 size, downscaled from 1024×1024 size, layer 5
```

illwieckz commented:
Well, if I do this (notice the `<< 3`):

```c++
	scalingStep = CheckImage2D( internalFormat, scaledWidth << 3, scaledHeight << 3, 0, format, GL_UNSIGNED_BYTE );
```

it works:

```
Warn: Image env/shared_space_src/sky07 too large, downscaling 1 time(s) to fit the hardware limits 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 0 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 1 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 2 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 3 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 4 
Debug: Uploading image env/shared_space_src/sky07 with 512×512 size, downscaled from 1024×1024 size, layer 5 
```

illwieckz commented:
I noticed that, on this computer reporting `1024` max texture size, if I multiply a `1024` sized texture by `4`, it does not scale down, then display garbage, meaning `4096` is expected to be OK but is not. While multiplying by `5` downscaling happens (and then displays the skybox properly).

illwieckz commented:
The current branch works, but I have no explanation for the multiplier. I've set `6` like the amount of faces a cube has, and a comment explaining there is no explanation.

illwieckz commented:
So, I renamed the `wat` commit to `upload image mips using mipWidth and mipHeight size`, that makes sense to me.

illwieckz commented:
At this point I think it's ready.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -785,6 +785,81 @@ R_ConvertBC5Image(const byte **in, byte **out, int numMips, int numLayers,
+	{
+		width >>= 1;
+		height >>= 1;
+		return 1 + CheckImage2D( internalFormat, width, height, border, format, type );
```
These functions need to return if width and height get down to 1, to make sure there can't be infinite recursion.

illwieckz replied:
Done

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -833,13 +896,61 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips,
 	scaledWidth = image->width;
 	scaledHeight = image->height;
 
+	int picMin = r_picMin->integer;
```
The logic to decide mip level is getting very long. Can you put it in a separate function?

illwieckz replied:
Done

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -796,6 +785,81 @@ R_ConvertBC5Image(const byte **in, byte **out, int numMips, int numLayers,
+	and if there is one, what is the multiplier to use,
+	so let's just use the amount of faces, 6. */
+
+	glTexImage2D( GL_PROXY_TEXTURE_CUBE_MAP, 0, internalFormat, width * 6, height * 6, border, format, type, nullptr );
```
This calculation doesn't seem right, since if you multiply both the width and height by 6, then the image uses 36 times as much memory.

But in any case it doesn't seem good to add a hack like this, since it could cause images to be wrongly downscaled on hardware that actually supports it.

illwieckz replied:
Deleted the bizarre `6`. Also, I noticed I was doing the check before internalFormat is determined (hence always using the same default one).

slipher commented:
Have you checked the OpenGL error codes to see if there's anything relevant in the cubemap bug case? I opened https://github.com/DaemonEngine/Daemon/pull/395 to enable this checking all the time.

illwieckz commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -811,13 +811,61 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips, image_t
+
+		while ( maxImageSize < imageSize )
+		{
+			imageSize >>= 2;
```
Right, I modified it to compare against width and height independently, like everything else in the existing code does. Also it looks like GL compare edges.

illwieckz commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -811,13 +811,61 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips, image_t
+		}
+
+		// Downscale image to picMax size.
+		int maxImageSize = picMax * picMax;
```
Modified it to compare against width and height independently.

illwieckz commented:
I made an extra function to downscale dimensions that not only applies on `scaleHeight` and `scaleWidth` but also on some other array and mip data. I noticed the extra computation was only done with `picmip` (and the variants I introduced), but not when the image is larger than the hardware can do.

We may technically do the custom downscale (picmip & all) and and the hardware-limited downscale at the same time after `internalFormat` is determined, and call the function to do the downscaling computation, but we have code to convert formats in the code determining internalFormat, so doing the custom downscale before determining the internal format may make conversion faster.

illwieckz commented:
> Have you checked the OpenGL error codes to see if there's anything relevant in the cubemap bug case? I opened #395 to enable this checking all the time.

If I run your PR (i.e. without my own patch from this PR), I get `Warn: OpenGL error GL_INVALID_ENUM detected` in `GLShaderManager::buildPermutation`, but no error when loading the skybox (that displays garbage).

illwieckz commented:
Except those two questions:

https://github.com/DaemonEngine/Daemon/pull/394/files#r519607355
https://github.com/DaemonEngine/Daemon/pull/394/files#r519510857

I consider this PR ready again.

Thanks to `GL_CheckErrors` I spotted some issues I introduced in my code, now fixed.

Note: The skybox issue is still not fixed but I give up on getting it fixed for 0.52. Anyway, hardware reproducing the issue runs better with `low` preset and this PR now introduces a nice feature that would make the game look better with `low` preset while workarounding the issue.

illwieckz commented:
Also, `r_picMin`, `r_picMax`, `picMin` and `picMax` keywords are temporary.

I'm looking for better cvar name than `r_picMin` and `r_picMax`, some base name that would fit well when ported to “dot” naming. Something like `maximumEdge` and `minimumEdge`, so we can do something like `renderer.textures.maximumEdge` or something like that. Material keywords would be edited accordingly.

Does anyone has an idea about that?

slipher commented on src/engine/renderer/InternalImage.h:
```diff
@@ -0,0 +1,48 @@
+#define INTERNAL_IMAGE_H
+#include "tr_local.h"
+
+int GetInternalImageSize( image_t *image );
```
Use `const image_t *` on all these

illwieckz replied:
Done, except on `R_GetImageHardwareScalingStep` which updates `*image`.

slipher commented:
For cvar naming how about `maxImageDimension`?

Did you ever find any cases where the texture proxy successfully identified a texture that couldn't be loaded?

illwieckz commented on src/engine/renderer/InternalImage.cpp:
```diff
@@ -0,0 +1,1007 @@
+			break;
+
+		default:
+			ASSERT( false );
```
I got this:

```
error: ‘ASSERT_UNREACHABLE’ was not declared in this scope
```

illwieckz replied:
Uh I feel stupid, there is no argument to `ASSERT_UNREACHABLE`… 🤦‍♀️

illwieckz commented on src/engine/renderer/InternalImage.cpp:
```diff
@@ -0,0 +1,1007 @@
+			}
+		}
+
+		scalingStep = std::max( scalingStep, std::max( 0, r_picmip->integer ) );
```
I think we better just remove the enforced range of 0 to 3, I see no reason for this enforced range, especially since we may reduce an image more than `r_picmip 3` would do (even, automatically when checking against hardware capabilities). `r_picmip 4` on a 8K texture is still huge. This range, like the whole `r_picmip` feature, assumes we use `128×128` textures everywhere.

illwieckz replied:
I added a commit to not enforce that range anymore.

illwieckz commented on src/engine/renderer/tr_local.h:
```diff
@@ -584,6 +584,14 @@ static inline void halfToFloat( const f16vec4_t in, vec4_t out )
 
+	struct imageParams_t
+	{
+		int bits;
```
Do I have something to do for types as well?

illwieckz commented on src/engine/renderer/InternalImage.cpp:
```diff
@@ -0,0 +1,1007 @@
+		// If image has picMax value set, enforce downscaling using this value.
+		if ( imagePicMax != 0 )
+		{
+			picMax = imagePicMax;
```
I added a test to not use the shader value when the cvar value is defined and smaller.

illwieckz commented on src/engine/renderer/InternalImage.cpp:
```diff
@@ -0,0 +1,1007 @@
+		image->uploadWidth >>= 1;
+		image->uploadHeight >>= 1;
+
+		if ( image->uploadWidth == 1 || image->uploadHeight == 1 )
```
done

illwieckz commented:
I did `imageMinDimension`/`imageMaxDimension`, `r_imageMinDimension`/`r_imageMaxDimension`.

illwieckz commented:
I got this on debug build, I don't know how to fix it:

```gdb
#0  0x00005555558d618e in Str::AssertOnTinyFormatError (reason="\"tinyformat: Not enough conversion specifiers in format string\"") at /home/illwieckz/dev/buildme-unvanquished/Daemon/src/common/String.cpp:37
#1  0x0000555555650b73 in tinyformat::detail::streamStateFromFormat (out=..., spacePadPositive=@0x7fffffff4181: false, ntrunc=@0x7fffffff4188: -1, fmtStart=0x5555559bcf38 "", formatters=0x7fffffff4410, argIndex=@0x7fffffff4184: 5, numFormatters=7) at /home/illwieckz/dev/buildme-unvanquished/Daemon/libs/tinyformat/tinyformat.h:607
#2  0x00005555556518e2 in tinyformat::detail::formatImpl (out=..., fmt=0x5555559bcf38 "", formatters=0x7fffffff4410, numFormatters=7) at /home/illwieckz/dev/buildme-unvanquished/Daemon/libs/tinyformat/tinyformat.h:807
#3  0x0000555555651da9 in tinyformat::vformat (out=..., fmt=0x5555559bcef8 "Uploading compressed image %s with %d×%d size, mip %d, layer %d", list=...) at /home/illwieckz/dev/buildme-unvanquished/Daemon/libs/tinyformat/tinyformat.h:963
#4  0x00005555558223c1 in tinyformat::format<char [64], int, int, unsigned short, unsigned short, int, int> (out=..., fmt=0x5555559bcef8 "Uploading compressed image %s with %d×%d size, mip %d, layer %d") at /home/illwieckz/dev/buildme-unvanquished/Daemon/libs/tinyformat/tinyformat.h:973
#5  0x0000555555821a7f in tinyformat::format<char [64], int, int, unsigned short, unsigned short, int, int> (fmt=0x5555559bcef8 "Uploading compressed image %s with %d×%d size, mip %d, layer %d") at /home/illwieckz/dev/buildme-unvanquished/Daemon/libs/tinyformat/tinyformat.h:982
#6  0x0000555555820e8b in Str::Format<char (&) [64], int&, int&, unsigned short&, unsigned short&, int&, int&> (format=...) at /home/illwieckz/dev/buildme-unvanquished/Daemon/src/common/String.h:315
#7  0x000055555581faa9 in Log::Logger::Debug<char (&) [64], int&, int&, unsigned short&, unsigned short&, int&, int&> (this=0x555556236c80 <Log::defaultLogger>, format=...) at /home/illwieckz/dev/buildme-unvanquished/Daemon/src/common/Log.h:220
#8  0x000055555581e932 in Log::Debug<char (&) [64], int&, int&, unsigned short&, unsigned short&, int&, int&> (format=...) at /home/illwieckz/dev/buildme-unvanquished/Daemon/src/common/Log.h:272
#9  0x00005555558189c2 in R_UploadImage (dataArray=0x7fffffff4b18, numLayers=1, numMips=8, image=0x7fffd5f3b200, imageParams=0x7fffffffcf70) at /home/illwieckz/dev/buildme-unvanquished/Daemon/src/engine/renderer/tr_image.cpp:1127
```

illwieckz commented:
I close this as commits will be cherry-picked in dedicated PR.
